infrastructure/src/main/java/org/corfudb/infrastructure/AutoCommitService.java

```diff
@@ -0,0 +1,217 @@
+                    initializeTails(currentLayout);
+                    return;
+                }
+
+                // Fetch maximum trim mark from log units and compare with committed tail.
+                // In order not to fetch trim mark for every retry, we only do this when
+                // the epoch changes after last attempt.
+                long currEpoch = currentLayout.getEpoch();
+                if (lastEpoch < 0 || trimmed || currEpoch != lastEpoch) {
+                    Token trimMark = getCorfuRuntime().getAddressSpaceView().getTrimMark(false);
+                    // Make sure all log units have same trim mark, then we can start committing
+                    // from the trim mark instead of the trailing committed tail.
+                    if (committedTail < trimMark.getSequence() - 1) {
+                        Token trimToken = new Token(trimMark.getEpoch(), trimMark.getSequence() - 1);
+                        getCorfuRuntime().getAddressSpaceView().prefixTrim(trimToken, false);
+                        committedTail = trimMark.getSequence() - 1;
+                    }
+                    lastEpoch = currEpoch;
+                    trimmed = false;
+                }
+
+                log.debug("runAutoCommit: trying to commit [{}, {}].", committedTail + 1, lastLogTail);
+
+                // Commit addresses in batches, retry limit is shared by all batches in this cycle.
+                // NOTE: This implementation relies on the fact that state transfer invokes the
+                // read protocol and fill holes, otherwise the committedTail could be invalid.
+                // (e.g. 1. State transferred a hole at address 100 from A to B; 2. Commit address
+                // 100, which only goes to A as B is not in this address segment; 3. B finishes
+                // state transfer and segments merged; 4. Send a new committedTail to B which is
+                // invalid as 100 is still a hole on B.
+                while (committedTail < lastLogTail) {
+                    long commitStart = committedTail + 1;
+                    long commitEnd = Math.min(committedTail + COMMIT_BATCH_SIZE, lastLogTail);
+                    getCorfuRuntime().getAddressSpaceView().commit(commitStart, commitEnd);
+                    log.trace("runAutoCommit: successfully committed batch [{}, {}]", committedTail, commitEnd);
+                    committedTail = commitEnd;
+                }
+
+                log.debug("runAutoCommit: successfully finished auto commit cycle. " +
+                        "New committed tail: {}.", committedTail);
+                break;
+
+            } catch (RuntimeException re) {
+                log.warn("runAutoCommit: encountered an exception on attempt {}/{}.",
+                        i, MAX_COMMIT_RETRY, re);
+
+                if (i >= MAX_COMMIT_RETRY) {
+                    log.warn("runAutoCommit: retry exhausted, abort and wait for next cycle");
```

Comment:
no need for this

Reply:
Since now we don't log on every exception, we need this.